File: lib/socialwise-flow/classification.ts
/**
 * Intelligent Intent Classification System for SocialWise Flow
 * Implements embedding-first classification with performance bands and degradation strategies
 * (versão multi-vetor: centroide + aliases do Redis, fallback DB)
 * 
 * Alteração solicitada:
 * - A banda LOW foi removida e substituída pela banda ROUTER (strategy: "router_llm").
 * - Quando score for insuficiente, ou não houver candidatos, ou ocorrer degradação/erro,
 *   roteamos para o Router LLM em vez de sugerir tópicos fixos.
 */

import { getPrismaInstance, getRedisInstance } from "@/lib/connections";
import { embeddingGenerator } from "@/lib/ai-integration/services/embedding-generator";
import { createLogger } from "@/lib/utils/logger";
import { IntentCandidate, AgentConfig } from "@/services/openai";

import {
  selectDegradationStrategy,
  shouldDegrade,
  determineFailurePoint,
  DegradationContext,
} from "./degradation-strategies";

// -----------------------------
// Config & Consts
// -----------------------------
const classificationLogger = createLogger("SocialWise-Classification");
const OPENAI_EMBED_MODEL = "text-embedding-3-small"; // 1536 dims

// Legal keywords that should force SOFT band even with low embedding scores
const LEGAL_KEYWORDS = [
  "mandado de segurança", "ms",
  "habeas corpus", "habeas data",
  "recurso", "recurso de multa", "multa de trânsito", "detran",
  "indenização", "ação judicial", "processo", "petição", "liminar",
  "direito", "advogado", "justiça", "tribunal", "juiz",
  "código", "lei", "constituição", "estatuto"
];

/**
 * Check if text contains legal keywords that should promote to SOFT band
 */
function checkLegalKeywords(text: string): boolean {
  const t = (text || "").toLowerCase();
  return LEGAL_KEYWORDS.some(k => t.includes(k));
}

// -----------------------------
// Types
// -----------------------------
export interface ClassificationResult {
  band: "HARD" | "SOFT" | "ROUTER";
  score: number;
  candidates: IntentCandidate[];
  strategy:
    | "direct_map"
    | "warmup_buttons"
    | "domain_topics"
    | "router_llm"
    | "direct_map_degraded"
    | "warmup_buttons_degraded"
    | "domain_topics_degraded"
    | "domain_topics_error";
  metrics: {
    embedding_ms?: number;
    llm_warmup_ms?: number;
    route_total_ms: number;
  };
}

export interface EmbeddingSearchResult {
  intent: string;
  score: number;
  description?: string;
  metadata?: Record<string, any>;
}

interface ScoredIntent {
  slug: string;
  name: string;
  desc: string;
  score: number;
  threshold: number;
  meta?: Record<string, any>;
}

// -----------------------------
// Math & Normalization helpers
// -----------------------------
function cosineSimilarity(a: number[], b: number[]): number {
  let dot = 0, na = 0, nb = 0;
  const len = Math.min(a.length, b.length);
  for (let i = 0; i < len; i++) {
    const x = a[i] || 0;
    const y = b[i] || 0;
    dot += x * y;
    na += x * x;
    nb += y * y;
  }
  if (na === 0 || nb === 0) return 0;
  return dot / (Math.sqrt(na) * Math.sqrt(nb));
}

function l2norm(v: number[]) { return Math.sqrt(v.reduce((s, x) => s + x * x, 0)) || 1; }
function l2normalize(v: number[]) { const n = l2norm(v); return v.map(x => x / n); }

/** Normalização idêntica à da rota (para casar vetores) */
function normalizeText(t: string) {
  return (t || "")
    .toLowerCase()
    .normalize("NFD").replace(/\p{Diacritic}/gu, "")
    .replace(/[^\p{L}\p{N}\s]/gu, " ")
    .replace(/\s+/g, " ")
    .trim();
}

// -----------------------------
// Embedding generation (HTTP direto, com timeout por chamada)
// -----------------------------
async function embedText(
  text: string,
  _timeoutMs = 1000 // mantemos a assinatura; o generator já tem timeout próprio
): Promise<number[] | null> {
  if (!process.env.OPENAI_API_KEY) return null;
  try {
    const emb = await embeddingGenerator.generateEmbedding(text, {
      normalize: true, trim: true, lowercase: true, removeExtraSpaces: true,
    });
    return l2normalize(emb.values); // ← HIT/MISS automaticamente cacheado no Redis
  } catch (error: any) {
    classificationLogger.warn("Embedding generation failed (cached path)", {
      error: error instanceof Error ? error.message : String(error),
    });
    return null;
  }
}

// -----------------------------
// Redis loaders (centroide + aliases) com fallback DB
// -----------------------------
type IntentRow = {
  id: string;
  name: string;
  description: string | null;
  similarityThreshold: number | null;
  embedding: number[] | null; // legado: centroide salvo no Postgres
};

async function loadIntentVectors(intent: IntentRow) {
  const redis = getRedisInstance();
  try {
    const key = `ai:intent:${intent.id}:emb`;
    const h = await redis.hgetall(key); // { model, centroid, aliases, aliases_text, updatedAt }
    if (h && h.centroid && h.aliases) {
      const centroid = JSON.parse(h.centroid) as number[];
      const aliases = JSON.parse(h.aliases) as number[][];
      const aliasTexts = h.aliases_text ? (JSON.parse(h.aliases_text) as string[]) : [];
      if (Array.isArray(centroid) && Array.isArray(aliases)) {
        return { centroid, aliases, aliasTexts, source: "redis" as const };
      }
    }
  } catch (e: any) {
    classificationLogger.warn("Failed to load vectors from Redis", {
      id: intent.id,
      err: e?.message || e,
    });
  }
  // Fallback para o vetor legado do DB (centroide)
  if (Array.isArray(intent.embedding) && intent.embedding.length) {
    return { centroid: intent.embedding as number[], aliases: [] as number[][], aliasTexts: [], source: "db" as const };
  }
  return null;
}

// -----------------------------
// Keyword fallback (mantido para degradação)
// -----------------------------
function performKeywordMatching(
  userText: string,
  intents: IntentRow[]
): IntentCandidate[] {
  const text = userText.toLowerCase();
  const matches: Array<{ intent: IntentRow; score: number }> = [];

  for (const intent of intents) {
    let score = 0;
    const name = (intent.name || "").toLowerCase();
    const desc = (intent.description || "").toLowerCase();

    const keywords = [...name.split(/\s+/), ...desc.split(/\s+/)].filter(k => k.length > 2);
    for (const keyword of keywords) {
      if (text.includes(keyword)) score += 0.1;
    }
    if (text.includes(name)) score += 0.3;

    if (score > 0) matches.push({ intent, score });
  }

  return matches
    .sort((a, b) => b.score - a.score)
    .slice(0, 3)
    .map((m) => ({
      slug: m.intent.name,
      name: m.intent.name,
      desc: m.intent.description || "",
      score: Math.min(m.score, 0.6),
    }));
}

// -----------------------------
// Embedding search (multi-vetor + degradação)
// -----------------------------
async function searchSimilarIntents(
  userText: string,
  userId: string,
  embeddingTimeoutMs?: number,
  context?: { channelType: string; inboxId: string; traceId?: string }
): Promise<{ candidates: IntentCandidate[]; searchMs: number; degraded?: boolean }> {
  const t0 = Date.now();
  const prisma = getPrismaInstance();

  try {
    // Intents ativas com *algum* embedding salvo (DB) para o usuário
    const intents: IntentRow[] = await (prisma as any).intent.findMany({
      where: { createdById: userId, isActive: true, embedding: { not: null } },
      select: { id: true, name: true, description: true, similarityThreshold: true, embedding: true },
    });

    if (!intents.length) {
      return { candidates: [], searchMs: Date.now() - t0 };
    }

    // 0) Alias direct hit: evita chamada de embedding quando a frase contém um alias
    const normText = normalizeText(userText);
    let bestHit: { intent: IntentRow; alias: string } | null = null;
    for (const it of intents) {
      const pack = await loadIntentVectors(it);
      if (!pack || !Array.isArray(pack.aliasTexts) || !pack.aliasTexts.length) continue;
      for (const a of pack.aliasTexts) {
        const na = normalizeText(a);
        if (na && normText.includes(na)) {
          if (!bestHit || na.length > normalizeText(bestHit.alias).length) {
            bestHit = { intent: it, alias: a };
          }
        }
      }
    }
    if (bestHit) {
      classificationLogger.info("Alias direct hit (no embedding call)", {
        intent: bestHit.intent.name,
        alias: bestHit.alias,
      });
      return {
        candidates: [{
          slug: bestHit.intent.name,
          name: bestHit.intent.name,
          desc: bestHit.intent.description || "",
          score: 0.95, // confiança alta por match textual
          threshold: typeof bestHit.intent.similarityThreshold === "number"
            ? bestHit.intent.similarityThreshold : 0.8,
        }],
        searchMs: Date.now() - t0,
        degraded: false,
      };
    }

    // 1) (somente se não bateu alias) gerar embedding da query
    const timeoutMs = embeddingTimeoutMs || 1500;
    const qVec = await embedText(userText, timeoutMs);

    if (!qVec) {
      classificationLogger.warn("Embedding generation failed, using keyword fallback", {
        userText: userText.substring(0, 50),
        userId,
        traceId: context?.traceId,
      });
      const keywordCandidates = performKeywordMatching(userText, intents);
      return { candidates: keywordCandidates, searchMs: Date.now() - t0, degraded: true };
    }

    const scored: ScoredIntent[] = [];
    for (const it of intents) {
      const pack = await loadIntentVectors(it);
      if (!pack) continue;

      const centroid = pack.centroid;
      const aliases = pack.aliases || [];

      let base = cosineSimilarity(qVec, centroid);
      let aliasMax = -Infinity;
      let aliasIdx = -1;

      for (let i = 0; i < aliases.length; i++) {
        const s = cosineSimilarity(qVec, aliases[i]);
        if (s > aliasMax) {
          aliasMax = s;
          aliasIdx = i;
        }
      }

      const score = Math.max(base, aliasMax);
      scored.push({
        slug: it.name,
        name: it.name,
        desc: it.description || "",
        score: isFinite(score) ? score : 0,
        threshold: typeof it.similarityThreshold === "number" ? it.similarityThreshold : 0.8,
        meta: {
          source: (pack as any).source,
          base,
          aliasMax: isFinite(aliasMax) ? aliasMax : undefined,
          aliasIdx: aliasIdx >= 0 ? aliasIdx : undefined,
        }
      });
    }

    scored.sort((a, b) => b.score - a.score);
    const searchMs = Date.now() - t0;

    if (scored.length) {
      const top = scored[0];
      classificationLogger.info("Embedding search completed", {
        userText: userText.substring(0, 50),
        candidatesFound: scored.length,
        topScore: top.score,
        searchMs,
      });
      classificationLogger.debug("Top candidate meta", top.meta);
    }

    const candidates: IntentCandidate[] = scored.slice(0, 5).map(s => ({
      slug: s.slug,
      name: s.name,
      desc: s.desc,
      score: s.score,
      threshold: s.threshold,
    }));

    return { candidates, searchMs };
  } catch (error) {
    classificationLogger.error("Embedding search failed", {
      error: error instanceof Error ? error.message : String(error),
      traceId: context?.traceId,
    });

    if (shouldDegrade(error) && context) {
      const degradationContext: DegradationContext = {
        userText,
        channelType: context.channelType,
        inboxId: context.inboxId,
        traceId: context.traceId,
        failurePoint: determineFailurePoint(error),
        originalError: error instanceof Error ? error : undefined,
      };
      classificationLogger.info("Applying degradation strategy for embedding search failure", {
        failurePoint: degradationContext.failurePoint,
        traceId: context.traceId,
      });
    }

    return { candidates: [], searchMs: Date.now() - t0, degraded: true };
  }
}

// -----------------------------
// Public API (bands + degradation)
// -----------------------------
export async function classifyIntentEmbeddingFirst(
  userText: string,
  userId: string,
  agent: AgentConfig,
  context?: { channelType: string; inboxId: string; traceId?: string }
): Promise<ClassificationResult> {
  const startTime = Date.now();

  try {
    // Step 1: Embedding search with degradation support
    const { candidates, searchMs, degraded } = await searchSimilarIntents(
      userText,
      userId,
      agent.warmupDeadlineMs, // usar timeout configurável do assistente
      context
    );

    // -----------------------------
    // Alteração: SEMPRE rotear para ROUTER quando não houver intenção forte
    // -----------------------------
    if (candidates.length === 0) {
      // Nenhum candidato → route to router (em vez de LOW/domain_topics)
      return {
        band: "ROUTER",
        score: 0,
        candidates: [],
        strategy: "router_llm",
        metrics: { embedding_ms: searchMs, route_total_ms: Date.now() - startTime },
      };
    }

    const topCandidate = candidates[0];
    const score = typeof topCandidate?.score === "number" ? topCandidate.score : 0;

    // Ajuste de thresholds conforme modo
    const scoreThresholds = degraded
      ? { hard: 0.5, soft: 0.3 }     // keyword fallback
      : { hard: 0.8, soft: 0.65 };   // embedding normal

    // HARD band permanece igual
    if (score >= scoreThresholds.hard) {
      return {
        band: "HARD",
        score,
        candidates: [topCandidate],
        strategy: degraded ? "direct_map_degraded" : "direct_map",
        metrics: { embedding_ms: searchMs, route_total_ms: Date.now() - startTime },
      };
    }

    // SOFT band (warmup buttons) permanece igual
    if (score >= scoreThresholds.soft) {
      return {
        band: "SOFT",
        score,
        candidates: candidates.slice(0, 3),
        strategy: degraded ? "warmup_buttons_degraded" : "warmup_buttons",
        metrics: { embedding_ms: searchMs, route_total_ms: Date.now() - startTime },
      };
    }

    // Promotion por palavras-chave legais continua existindo
    const hasLegalKeywords = checkLegalKeywords(userText);
    if (hasLegalKeywords && score >= 0.4) {
      classificationLogger.info("Legal keywords detected - promoting to SOFT band", {
        userText: userText.substring(0, 50),
        score,
        traceId: context?.traceId
      });
      return {
        band: "SOFT",
        score: Math.max(score, 0.65),
        candidates: candidates.slice(0, 3),
        strategy: degraded ? "warmup_buttons_degraded" : "warmup_buttons",
        metrics: { embedding_ms: searchMs, route_total_ms: Date.now() - startTime },
      };
    }

    // Caso contrário → ROUTER (substitui LOW)
    return {
      band: "ROUTER",
      score,
      candidates: candidates.slice(0, 3), // passa hints para o Router LLM se desejarem
      strategy: "router_llm",
      metrics: { embedding_ms: searchMs, route_total_ms: Date.now() - startTime },
    };
  } catch (error) {
    classificationLogger.error("Classification failed", {
      error: error instanceof Error ? error.message : String(error),
      traceId: context?.traceId,
    });

    if (shouldDegrade(error) && context) {
      const degradationContext: DegradationContext = {
        userText,
        channelType: context.channelType,
        inboxId: context.inboxId,
        traceId: context.traceId,
        failurePoint: determineFailurePoint(error),
        originalError: error instanceof Error ? error : undefined,
      };
      classificationLogger.info("Classification failed, routing to Router LLM", {
        failurePoint: degradationContext.failurePoint,
        traceId: context.traceId,
      });
    }

    // Em erro → também ROUTER
    return {
      band: "ROUTER",
      score: 0,
      candidates: [],
      strategy: "router_llm",
      metrics: { embedding_ms: 0, route_total_ms: Date.now() - startTime },
    };
  }
}

/**
 * Router LLM classification for embedipreview=false mode
 * (Mantido; pode ser usado quando você quiser bypassar embeddings)
 */
export async function classifyIntentRouterLLM(
  userText: string,
  agent: AgentConfig
): Promise<ClassificationResult> {
  const startTime = Date.now();
  return {
    band: "ROUTER",
    score: 1.0,
    candidates: [],
    strategy: "router_llm",
    metrics: { route_total_ms: Date.now() - startTime },
  };
}

/**
 * Main classification entry point with degradation support
 */
export async function classifyIntent(
  userText: string,
  userId: string,
  agent: AgentConfig,
  embedipreview = true,
  context?: { channelType: string; inboxId: string; traceId?: string }
): Promise<ClassificationResult> {
  if (!embedipreview) {
    return classifyIntentRouterLLM(userText, agent);
  }
  return classifyIntentEmbeddingFirst(userText, userId, agent, context);
}

File: platform-backend/domains/socialwise/services/intent/classification.py
"""Embedding-first intent classification for SocialWise Flow."""

from __future__ import annotations

import math
import unicodedata
from dataclasses import dataclass
from time import perf_counter

from redis.asyncio import Redis
from sqlalchemy import text

from domains.socialwise.db.session_compat import session_ctx
from domains.socialwise.services.intent.cache import SocialwiseIntentCache
from domains.socialwise.services.intent.types import (
    AssistantConfig,
    CacheKeyConfig,
    ClassificationMetrics,
    ClassificationResult,
    IntentCandidate,
)
from platform_core.ai.litellm_config import call_embedding, resolve_litellm_model
from platform_core.config import settings
from platform_core.logging.config import get_logger

logger = get_logger(__name__)

LEGAL_KEYWORDS = [
    "mandado de segurança",
    "habeas corpus",
    "habeas data",
    "recurso",
    "multa de trânsito",
    "detran",
    "indenização",
    "ação judicial",
    "processo",
    "petição",
    "liminar",
    "direito",
    "advogado",
    "justiça",
    "tribunal",
    "juiz",
    "lei",
    "constituição",
    "estatuto",
]


@dataclass(slots=True)
class IntentRow:
    id: str
    name: str
    slug: str
    description: str | None
    similarity_threshold: float | None
    embedding: list[float] | None


@dataclass(slots=True)
class IntentVectorPack:
    centroid: list[float]
    aliases: list[list[float]]
    alias_texts: list[str]
    source: str


def _normalize_text(value: str) -> str:
    normalized = unicodedata.normalize("NFD", value or "")
    without_marks = "".join(ch for ch in normalized if unicodedata.category(ch) != "Mn")
    filtered = "".join(ch if ch.isalnum() or ch.isspace() else " " for ch in without_marks.lower())
    return " ".join(filtered.split())


def _l2normalize(vector: list[float]) -> list[float]:
    norm = math.sqrt(sum(item * item for item in vector)) or 1.0
    return [item / norm for item in vector]


def _cosine_similarity(a: list[float], b: list[float]) -> float:
    if not a or not b:
        return 0.0
    length = min(len(a), len(b))
    dot = sum((a[idx] or 0.0) * (b[idx] or 0.0) for idx in range(length))
    norm_a = math.sqrt(sum((a[idx] or 0.0) ** 2 for idx in range(length)))
    norm_b = math.sqrt(sum((b[idx] or 0.0) ** 2 for idx in range(length)))
    if norm_a == 0 or norm_b == 0:
        return 0.0
    return dot / (norm_a * norm_b)


def _parse_vector(raw_value: str | None) -> list[float] | None:
    if not raw_value:
        return None
    cleaned = raw_value.strip().removeprefix("[").removesuffix("]")
    if not cleaned:
        return None
    try:
        return [float(chunk.strip()) for chunk in cleaned.split(",") if chunk.strip()]
    except ValueError:
        return None


def _has_legal_keywords(text_value: str) -> bool:
    lowered = (text_value or "").lower()
    return any(keyword in lowered for keyword in LEGAL_KEYWORDS)


def _perform_keyword_matching(user_text: str, intents: list[IntentRow]) -> list[IntentCandidate]:
    lowered = (user_text or "").lower()
    matches: list[IntentCandidate] = []
    for intent in intents:
        score = 0.0
        name = (intent.name or "").lower()
        description = (intent.description or "").lower()
        keywords = [token for token in f"{name} {description}".split() if len(token) > 2]
        for keyword in keywords:
            if keyword in lowered:
                score += 0.1
        if name and name in lowered:
            score += 0.3
        if score > 0:
            matches.append(
                IntentCandidate(
                    slug=intent.slug,
                    name=intent.name,
                    desc=intent.description or "",
                    score=min(score, 0.6),
                    threshold=intent.similarity_threshold or 0.8,
                )
            )
    matches.sort(key=lambda item: item.score or 0, reverse=True)
    return matches[:3]


async def _load_active_intents(user_id: str) -> list[IntentRow]:
    async with session_ctx() as session:
        result = await session.execute(
            text(
                """
                SELECT
                    id,
                    name,
                    slug,
                    description,
                    "similarityThreshold" AS similarity_threshold,
                    embedding::TEXT AS embedding
                FROM "Intent"
                WHERE "createdById" = :user_id
                  AND "isActive" = true
                """
            ),
            {"user_id": user_id},
        )
        rows = result.mappings().all()

    return [
        IntentRow(
            id=str(row["id"]),
            name=str(row["name"]),
            slug=str(row["slug"]),
            description=row["description"],
            similarity_threshold=float(row["similarity_threshold"] or 0.8),
            embedding=_parse_vector(row["embedding"]),
        )
        for row in rows
    ]


async def _load_vector_pack(redis_client: Redis, intent: IntentRow) -> IntentVectorPack | None:
    try:
        raw_hash = await redis_client.hgetall(f"ai:intent:{intent.id}:emb")
        if raw_hash and raw_hash.get("centroid"):
            centroid = raw_hash.get("centroid")
            aliases = raw_hash.get("aliases")
            alias_texts = raw_hash.get("aliases_text")
            parsed_centroid = _parse_vector(centroid)
            parsed_aliases = []
            if aliases:
                import json

                parsed_aliases_raw = json.loads(aliases)
                if isinstance(parsed_aliases_raw, list):
                    parsed_aliases = [
                        _l2normalize([float(value) for value in vector])
                        for vector in parsed_aliases_raw
                        if isinstance(vector, list)
                    ]
            parsed_alias_texts: list[str] = []
            if alias_texts:
                import json

                loaded_alias_texts = json.loads(alias_texts)
                if isinstance(loaded_alias_texts, list):
                    parsed_alias_texts = [str(item) for item in loaded_alias_texts]
            if parsed_centroid:
                return IntentVectorPack(
                    centroid=_l2normalize(parsed_centroid),
                    aliases=parsed_aliases,
                    alias_texts=parsed_alias_texts,
                    source="redis",
                )
    except Exception as exc:
        logger.warning("intent_vector_pack_redis_error", intent_id=intent.id, error=str(exc))

    if intent.embedding:
        return IntentVectorPack(
            centroid=_l2normalize(intent.embedding),
            aliases=[],
            alias_texts=[],
            source="db",
        )
    return None


async def _generate_query_embedding(
    user_text: str,
    cache_config: CacheKeyConfig | None,
    cache: SocialwiseIntentCache | None,
) -> list[float] | None:
    if cache and cache_config:
        cached = await cache.get_embedding(cache_config, user_text)
        if cached:
            return cached

    model = resolve_litellm_model("openai", settings.openai_embedding_model)
    try:
        response = await call_embedding(
            model,
            user_text,
            timeout=10,
            retries=2,
            base_delay_ms=500,
            max_delay_ms=2000,
        )
    except Exception as exc:
        logger.warning("intent_query_embedding_error", error=str(exc))
        return None

    vector = response.vectors[0] if response.vectors else None
    if not vector:
        return None
    normalized = _l2normalize(vector)
    if cache and cache_config:
        await cache.set_embedding(cache_config, user_text, normalized)
    return normalized


async def classify_intent(
    user_text: str,
    user_id: str,
    agent: AssistantConfig,
    *,
    embedipreview: bool = True,
    channel_type: str = "whatsapp",
    inbox_id: str = "",
    trace_id: str | None = None,
    cache_config: CacheKeyConfig | None = None,
    cache: SocialwiseIntentCache | None = None,
) -> ClassificationResult:
    start = perf_counter()
    cache_client = cache or SocialwiseIntentCache()

    if cache_config:
        cached = await cache_client.get_classification_result(cache_config, user_text)
        if cached:
            return cached

    if not embedipreview:
        result = ClassificationResult(
            band="ROUTER",
            score=1.0,
            candidates=[],
            strategy="router_llm_bypass",
            metrics=ClassificationMetrics(route_total_ms=0),
        )
        if cache_config:
            await cache_client.set_classification_result(cache_config, user_text, result)
        return result

    intents = await _load_active_intents(user_id)
    if not intents:
        result = ClassificationResult(
            band="ROUTER",
            score=0.0,
            candidates=[],
            strategy="router_llm",
            metrics=ClassificationMetrics(route_total_ms=int((perf_counter() - start) * 1000)),
        )
        if cache_config:
            await cache_client.set_classification_result(cache_config, user_text, result)
        return result

    redis_client = Redis.from_url(str(settings.redis_url), decode_responses=True)
    packs: dict[str, IntentVectorPack] = {}
    normalized_text = _normalize_text(user_text)
    best_alias_hit: tuple[IntentRow, str] | None = None

    for intent in intents:
        pack = await _load_vector_pack(redis_client, intent)
        if not pack:
            continue
        packs[intent.id] = pack
        for alias in pack.alias_texts:
            normalized_alias = _normalize_text(alias)
            if normalized_alias and normalized_alias in normalized_text:
                if best_alias_hit is None or len(normalized_alias) > len(_normalize_text(best_alias_hit[1])):
                    best_alias_hit = (intent, alias)

    if best_alias_hit:
        hit_intent, alias = best_alias_hit
        result = ClassificationResult(
            band="HARD",
            score=0.95,
            candidates=[
                IntentCandidate(
                    slug=hit_intent.slug,
                    name=hit_intent.name,
                    desc=hit_intent.description or "",
                    score=0.95,
                    threshold=hit_intent.similarity_threshold or 0.8,
                    aliases=packs.get(hit_intent.id, IntentVectorPack([], [], [], "db")).alias_texts,
                    alias_matched=alias,
                )
            ],
            strategy="direct_map",
            metrics=ClassificationMetrics(
                route_total_ms=int((perf_counter() - start) * 1000),
                embedding_ms=0,
            ),
        )
        if cache_config:
            await cache_client.set_classification_result(cache_config, user_text, result)
        return result

    embedding_started = perf_counter()
    query_vector = await _generate_query_embedding(user_text, cache_config, cache_client if cache_config else None)
    embedding_ms = int((perf_counter() - embedding_started) * 1000)
    degraded = query_vector is None

    candidates: list[IntentCandidate] = []
    if query_vector is None:
        candidates = _perform_keyword_matching(user_text, intents)
    else:
        for intent in intents:
            pack = packs.get(intent.id)
            if not pack:
                pack = await _load_vector_pack(redis_client, intent)
                if not pack:
                    continue
                packs[intent.id] = pack

            centroid_score = _cosine_similarity(query_vector, pack.centroid)
            alias_scores = [_cosine_similarity(query_vector, alias_vector) for alias_vector in pack.aliases]
            best_alias_score = max(alias_scores) if alias_scores else float("-inf")
            score = max(centroid_score, best_alias_score)
            alias_index = alias_scores.index(best_alias_score) if alias_scores else -1
            alias_matched = pack.alias_texts[alias_index] if alias_index >= 0 and alias_index < len(pack.alias_texts) else None
            candidates.append(
                IntentCandidate(
                    slug=intent.slug,
                    name=intent.name,
                    desc=intent.description or "",
                    score=score if math.isfinite(score) else 0.0,
                    threshold=intent.similarity_threshold or 0.8,
                    aliases=pack.alias_texts,
                    alias_matched=alias_matched,
                )
            )
        candidates.sort(key=lambda item: item.score or 0, reverse=True)
        candidates = candidates[:5]

    score = (candidates[0].score or 0.0) if candidates else 0.0
    hard_threshold = 0.5 if degraded else 0.8
    soft_threshold = 0.3 if degraded else 0.65

    if not candidates:
        band = "ROUTER"
        strategy = "router_llm"
    elif score >= hard_threshold:
        band = "HARD"
        strategy = "direct_map_degraded" if degraded else "direct_map"
        candidates = candidates[:1]
    elif score >= soft_threshold or (_has_legal_keywords(user_text) and score >= 0.4):
        band = "SOFT"
        strategy = "warmup_buttons_degraded" if degraded else "warmup_buttons"
        candidates = candidates[:3]
    else:
        band = "ROUTER"
        strategy = "router_llm"
        candidates = candidates[:3]

    result = ClassificationResult(
        band=band,
        score=score,
        candidates=candidates,
        strategy=strategy,
        metrics=ClassificationMetrics(
            route_total_ms=int((perf_counter() - start) * 1000),
            embedding_ms=embedding_ms,
        ),
    )
    logger.info(
        "intent_classification_complete",
        user_id=user_id,
        inbox_id=inbox_id,
        channel_type=channel_type,
        trace_id=trace_id,
        band=result.band,
        strategy=result.strategy,
        score=score,
        degraded=degraded,
        candidates=[candidate.slug for candidate in result.candidates],
    )
    if cache_config:
        await cache_client.set_classification_result(cache_config, user_text, result)
    return result
